walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -17,59 +17,91 @@
 import java.util.ArrayList;
 import java.util.Collection;
 import java.util.Collections;
+import java.util.HashMap;
 import java.util.HashSet;
 import java.util.List;
 import java.util.Set;
 
 public final class FindMeetingQuery {
+  private String MANDATORY = "mandatory";
+  private String OPTIONAL = "optional";
+
   /**Return collection of timeRanges that work given the events all invited attendees are going
    * to.*/
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    Collection<TimeRange> withOptionalAttendees = findAvailableTimes(events, request, true);
-    Collection<String> mandatoryAttendees = request.getAttendees();
-    // If there are no mandatory attendees but all optional attendees, return requests
-    // as if all optional are mandatory
-    if (!mandatoryAttendees.isEmpty() && withOptionalAttendees.isEmpty()) {
-      Collection<TimeRange> withoutOptionalAttendees = findAvailableTimes(events, request, false);
-      return withoutOptionalAttendees;
-    } else {
-      return withOptionalAttendees;
-    }
-  }
-
-  private Collection<TimeRange> findAvailableTimes(
-      Collection<Event> events, MeetingRequest request, boolean withOptionalAttendees) {
     List<Event> eventsList = new ArrayList<Event>(events);
     Collections.sort(eventsList, Event.ORDER_BY_START);
     eventsList.add(null); // Run for loop additional time to account for EOD
+    long duration = request.getDuration();
 
-    List<TimeRange> availableTimes = new ArrayList<TimeRange>();
-    Set<String> requestAttendees = new HashSet<String>(request.getAttendees());
-    if (withOptionalAttendees) {
-      requestAttendees.addAll(request.getOptionalAttendees());
-    }
+    List<TimeRange> mandatoryTimes = new ArrayList<TimeRange>();
+    List<TimeRange> optionalTimes = new ArrayList<TimeRange>();
+
+    Set<String> mandatoryAttendees = new HashSet<String>(request.getAttendees());
+    Set<String> optionalAndMandatoryAttendees = new HashSet<String>(request.getOptionalAttendees());
+    optionalAndMandatoryAttendees.addAll(mandatoryAttendees);
 
-    TimeRange prevEventTime = null;
-    TimeRange currEventTime = null;
+    HashMap<String, TimeRange> prevTime = new HashMap<String, TimeRange>();
+    prevTime.put(MANDATORY, null);
+    prevTime.put(OPTIONAL, null);
+    TimeRange currTime = null;
+    boolean addToMandatory = true;
+    boolean addToOptional = true;
     for (Event currEvent : eventsList) {
-      currEventTime = null;
+      currTime = null;
+      addToMandatory = true;
+      addToOptional = true;
       if (currEvent != null) {
```

Comment:
Prefer ```if (currEvent == null) { continue; } // rest of the code to do work``` so that the rest of the code (which may have deeper nested structures) can live outside the ```if {}```. Linux code style actually forbids more than 3 layers' nested structure.